### ISSUE TITLE
0.3.5 release planning issue

### ISSUE THREAD

**Issue by staticfloat** [JuliaLang/julia#9631]
Due to #9513, we've decided to push the 0.3.5 release up two weeks, shooting for this weekend for a new release.  Let's discuss here what we want to try and include in this release.  Although it is motivated by a single issue, I see no reason to avoid other fixes that have already been merged into `release-0.3`/are close to being merged.  In short, this should be a fully-fledged release, not a targeted bugfix.

**Comment by tkelman** [comment]
There are roughly 25 issues or PR's currently labeled backport pending: https://github.com/JuliaLang/julia/issues?q=label%3A%22backport+pending%22+sort%3Aupdated-desc

Due to the abbreviated schedule we may want to make individual go/no-go decisions on backporting them into 0.3.5. Many are documentation-related, but not all.

Additional commits that were marked `@juliabackports` but don't have corresponding issues to apply the label to:
[ ] c4a81f11c5c69ba99aff0c7468d5d789a9a3325b
[ ] the documentation parts of 68a09e0f571320f96bcfb8affa2d74864ca514c9 that apply to release-0.3

**Comment by tkelman** [comment]
For help with backporting, here's my preferred process:

1. Look over commit, see if any pieces of it are specific to 0.4 (`UInt`, `AbstractString`, etc)
2. Try `git cherry-pick -x -e <sha>` to see if it applies cleanly. The `-x` is nice since it adds a cross-reference in the commit message, `-e` is useful to edit the message to link to whatever PR or issue # it came from, or add [ci skip] for anything that's docs-only.
3. Resolve conflicts if any. For really nasty conflicts you may have to do the backport by hand. For the doc reorganization I had to turn a few changes into patch files, manually tweaking the patch file to move some hunks to refer to different files, cut out a line or two of context, etc before it would `git apply`
4. `make test` locally
5. (optional) Push to a new personal branch if you want to check Travis on it. If you want AppVeyor to also test a commit before pushing it to release-0.3, you can either temporarily add the branch name to the whitelist in appveyor.yml, or make a PR for it against release-0.3.
5. Push to release-0.3
6. Revert if CI shows any accidental breakage caused by the backport
7. Leave a link in the corresponding issue/PR/commit comment linking to the backported sha
8. Remove backport pending label

**Comment by JeffBezanson** [comment]
I would be very conservative here. It would be quite unfortunate if a release to fix a regression itself had another regression.

**Comment by tkelman** [comment]
Okay, I got almost all of the obvious ones. Can someone else please

A) check my work
and
B) take care of (probably in this order): the documentation parts of 68a09e0 that apply to release-0.3, then manually do the equivalent of f910cd04c96b55b702c474811809e03d613f8235, then the equivalent of 0b5811d632a66751dfbce77865451ffd06c5d774

**Comment by tkelman** [comment]
Whoops, sorry, apparently gcd isn't type-stable on release-0.3 for 64 bit machines, my local tests were on win32. I'll have to comment out some of the backported tests that fail on 0.3.

**Comment by staticfloat** [comment]
* Documentation part of 68a09e0f571320f96bcfb8affa2d74864ca514c9 backported in c95156477a5e230dba0ebf1ab52dfa4087b5a532

* Profiling docs reorganization from f910cd04c96b55b702c474811809e03d613f8235 backported in 0a7393355197f85145afbb8df8d702355895c8f3

* stdlib/profile.rst cleanup from 0b5811d632a66751dfbce77865451ffd06c5d774 backported in 75df2f10d76d79ca29d04987f38b328baa476a4e

**Comment by staticfloat** [comment]
I'm going to run a `make release-candidate` on my machine, we'll see how it goes.

**Comment by tkelman** [comment]
Last doc change that might be ready would be `git diff 76c991d7~ 76c991d7 | sed -e 's/nothing/false/g' | git apply && git add doc/manual/control-flow.rst && git commit -C 76c991d7 && git commit --amend` (then add sha x-ref) from #9320.

**Comment by jiahao** [comment]
Would be nice to squeeze in an update to the Unicode table after #9584; simply run `doc/tabcomplete.jl` and pipe its output to `doc/manual/unicode-input-table.rst`. On current master, the latter currently does not compile due to a 'malformed table' error.

**Comment by staticfloat** [comment]
Also, we have a bit of a problem with sphinx; since it doesn't understand multiple dispatch, it's mad that we have a `Base.length()` defined in `arrays.rst` as well as `collections.rst`.  Thoughts on how to fix this?  Should we link to just one?

**Comment by jiahao** [comment]
I say ignore the warning; despite it, Sphinx generates cross-references to duplicated functions in the index (look at `_build/html/genindex.html`), which actually works nicely in our favor.

**Comment by tkelman** [comment]
@jiahao I did partially backport the documentation of the current status of unicode tab completion support, and regenerated the unicode-input-table.rst in that backport, but #9584 and an earlier one of @stevengj's PR's that added a large number of unicode symbols were not yet backported. Now might not be the best time to do so.

**Comment by jiahao** [comment]
@tkelman you can always run `tabcomplete.jl` on the release branch separately just to make sure the table builds.

**Comment by staticfloat** [comment]
Okay, weird.  I ran it a second time and it passed.  I think I must have been confusing two separate things here.

`make release-candidate` passes on my machine now, so that's good.

**Comment by tkelman** [comment]
Nobody objected to that last doc change (https://github.com/JuliaLang/julia/issues/9631#issuecomment-68942106) so I just backported it. I think we should be okay to tag now?

**Comment by staticfloat** [comment]
Yeah, I think we're good here as well.  I'll run `release-candidate` once more to be certain and then we can tag it and bag (err...build) it.

**Comment by tkelman** [comment]
pause. need to backport a bugfix that tim just committed to master

**Comment by staticfloat** [comment]
I also found a doctest failure.

**Comment by tkelman** [comment]
got the bugfix in 2d894e5d20d0df6c420abf3aa89b1bd22b6b25d4, carry on

**Comment by ivarne** [comment]
Can we get https://github.com/JuliaLang/julia/commit/23424b754fffc5967503b137e7f1b61c3775d38a as well. It seems relevant,  but I have not tested it.

**Comment by staticfloat** [comment]
Yes, I agree that 23424b7 looks good to have, especially given that this release is partly due to embedding concerns.  I'll grab that in as well.

**Comment by staticfloat** [comment]
Tagged in a05f87b79a.  Binaries building now.  @nalimilan I don't know if you've been following this at all, but it's time to work your RPM magic.

**Comment by waTeim** [comment]
Tested against the Linux and jl_ctx_base error is fixed but now there is a new one 

    fatal: error thrown and no exception handler available.
    ErrorException("Module: too many arguments (expected 1)")

about halfway through my tests.

Not sure if this is my error yet caused by the mod i made for my own stuff in #9661, have not debugged thoroughly, is this the right thread for this?

**Comment by waTeim** [comment]
That error was due to my workaround, so with the fix for #9661, in place I reverted, and all tests pass.  So I've verified at least as far as my stuff is concerned.

* 3.5 (Linux) is fine as is
* jl_new_module is usable in 0.4, and it's probably a good idea to back-port like you're already doing.

I'll test OS/X as it becomes available but I'm confident it will pass too.

**Comment by tkelman** [comment]
If you could provide reproduction steps that might help. Looking at the Travis results on your node-julia repo it looks like @staticfloat may still need to update the http://status.julialang.org/stable links, at least for OSX?

**Comment by nalimilan** [comment]
RPMs are OK now.
